internal/controller/pkg/revision/establisher.go

```diff
@@ -137,39 +137,64 @@ func (e *APIEstablisher) ReleaseObjects(ctx context.Context, parent v1.PackageRe
 	// referenced resources available and rebuilding the status.
 	// In the next reconciliation loop, and we will be able to remove the
 	// control/ownership of the objects using the new status.
-	objRefs := parent.GetObjects()
-	// Stop controlling objects.
-	for _, ref := range objRefs {
-		u := unstructured.Unstructured{}
-		u.SetAPIVersion(ref.APIVersion)
-		u.SetKind(ref.Kind)
-		u.SetName(ref.Name)
-
-		if err := e.client.Get(ctx, types.NamespacedName{Name: u.GetName()}, &u); err != nil {
-			if kerrors.IsNotFound(err) {
-				// This is not expected, but still not an error for relinquishing.
-				continue
-			}
-			return errors.Wrapf(err, errFmtGetOwnedObject, u.GetKind(), u.GetName())
+	allObjs := parent.GetObjects()
+	if len(allObjs) == 0 {
+		return nil
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(maxConcurrentEstablishers)
+	for _, ref := range allObjs {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
```

Comment:
maybe just break to wait for the other go routines? `g.Wait` below will return the first error of some go routine.

Reply:
but then we need another such context check below.

Also note: is `ctx.Err()` really the go routine error? There is a `context.Cause` function for that.

Reply:
Yes, looking into the context code, there is `err` ("context cancelled") and `cause` (the go routine error).

Reply:
https://github.com/crossplane/crossplane/pull/5044#discussion_r1400388847